framework/src/kernels/ADCoupledForce.C

```diff
@@ -0,0 +1,42 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADCoupledForce.h"
+
+registerMooseObject("MooseApp", ADCoupledForce);
+
+defineLegacyParams(ADCoupledForce);
+
+InputParameters
+ADCoupledForce::validParams()
+{
+  InputParameters params = ADKernel::validParams();
+
+  params.addClassDescription("Implements a source term proportional to the value of a coupled "
+                             "variable. Weak form: $(\\psi_i, -\\sigma v)$.");
+  params.addRequiredCoupledVar("v", "The coupled variable which provides the force");
+  params.addParam<Real>(
+      "coef", 1.0, "Coefficent ($\\sigma$) multiplier for the coupled force term.");
+
+  return params;
+}
+
+ADCoupledForce::ADCoupledForce(const InputParameters & parameters)
+  : ADKernel(parameters), _v_var(coupled("v")), _v(coupledValue("v")), _coef(getParam<Real>("coef"))
+{
+  if (_var.number() == _v_var)
+    mooseError("Coupled variable 'v' needs to be different from 'variable' with ADCoupledForce, "
+               "consider using Reaction or somethig similar");
```

Comment:
We could probably remove this restriction, in which case there's no need at all for the `_v_var` member

Reply:
I think when people do this, it actually indicates they have a problem in their input file, so we should tell them.

Reply:
Plus, I think we want to stay consistent between non-AD and AD.

Reply:
sounds good to me